File: 826/most-profit-assigning-work.go
package main

import (
	"fmt"
	"sort"
)

func maxProfitAssignment(difficulty []int, profit []int, worker []int) int {
	jobs := make([]job, len(difficulty))
	for i := range difficulty {
		jobs[i] = job{
			d: difficulty[i],
			p: profit[i],
		}
	}

	sort.Slice(jobs, func(i, j int) bool {
		return jobs[i].d < jobs[j].d
	})

	sort.Ints(worker)

	res := 0
	i := 0
	maxp := 0
	for _, ability := range worker {
		for i < len(jobs) && ability >= jobs[i].d {
			maxp = max(maxp, jobs[i].p)
			i++
		}
		res += maxp
	}

	return res
}

type job struct {
	d, p int
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	// difficulty = [2,4,6,8,10], profit = [10,20,30,40,50], worker = [4,5,6,7]
	difficulty := []int{2, 4, 6, 8, 10}
	profit := []int{10, 20, 30, 40, 50}
	worker := []int{4, 5, 6, 7}
	fmt.Println(maxProfitAssignment(difficulty, profit, worker))
}

File: 826/most-profit-assigning-work.py
class Solution:
    def maxProfitAssignment(self, difficulty, profit, worker):
        """
        :type difficulty: List[int]
        :type profit: List[int]
        :type worker: List[int]
        :rtype: int
        """
        from collections import defaultdict
        diffPro = defaultdict(int)
        for diff, pro in zip(difficulty, profit):
            diffPro[diff] = max(diffPro[diff], pro)
        maxVal = 0
        for x in range(min(difficulty + worker), max(difficulty + worker) + 1):
            diffPro[x] = max(diffPro[x], maxVal)
            maxVal = max(diffPro[x], maxVal)
        return sum(diffPro[w] for w in worker)
